src/snapred/backend/recipe/algorithm/FocusSpectraAlgorithm.py

```diff
@@ -0,0 +1,124 @@
+import json
+from ast import In
+from typing import Dict
+
+from mantid.api import (
+    AlgorithmFactory,
+    MatrixWorkspaceProperty,
+    PropertyMode,
+    PythonAlgorithm,
+)
+from mantid.kernel import Direction
+
+from snapred.backend.dao.ingredients.ReductionIngredients import ReductionIngredients as Ingredients
+from snapred.backend.dao.state.PixelGroup import PixelGroup
+from snapred.backend.recipe.algorithm.MantidSnapper import MantidSnapper
+
+
+class FocusSpectraAlgorithm(PythonAlgorithm):
+    def category(self):
+        return "SNAPRed Data Processing"
+
+    def PyInit(self):
+        # declare properties
+        self.declareProperty(
+            MatrixWorkspaceProperty("InputWorkspace", "", Direction.Input, PropertyMode.Mandatory),
+            doc="Workspace containing values at each pixel",
+        )
+        self.declareProperty(
+            MatrixWorkspaceProperty("GroupingWorkspace", "", Direction.Input, PropertyMode.Mandatory),
+            doc="Workspace defining the grouping for diffraction focusing",
+        )
+        self.declareProperty(
+            MatrixWorkspaceProperty("OutputWorkspace", "", Direction.Output, PropertyMode.Optional),
+            doc="The diffraction-focused data",
+        )
+        self.declareProperty("Ingredients", defaultValue="", direction=Direction.Input)
```

Comment:
can you add `optional=PropertyMode.Mandatory`?

Reply:
I think I have to say no to that, else it requires I input a string that actually refers to an existing workspace?

Reply:
sorry, I meant on the ingredients